Content.Server/Station/Systems/StationJobsSystem.Roundstart.cs

```diff
@@ -0,0 +1,319 @@
+                {
+                    var slots = stationSlots[station];
+
+                    // Get all of the jobs in the selected weight category.
+                    foreach (var (job, slot) in stationJobs[station])
+                    {
+                        if (_jobsByWeight[weight].Contains(job))
+                            slots.Add(job, slot);
+                    }
+                }
+
+                // Intentionally discounts the value of uncapped slots! They're only a single slot when deciding a station's share.
+                // Clear for reuse.
+                stationTotalSlots.Clear();
+                foreach (var (station, jobs) in stationSlots)
+                {
+                    stationTotalSlots.Add(
+                        station,
+                        (int)jobs.Values.Sum(x => x ?? 1)
+                        );
+                }
+
+                var totalSlots = 0;
+
+                foreach (var (_, slot) in stationTotalSlots)
+                {
+                    totalSlots += slot;
+                }
+
+                if (totalSlots == 0)
+                    continue; // No slots so just leave.
+
+                // Clear for reuse.
+                stationShares.Clear();
+                var distributed = 0;
+
+                foreach (var station in stations)
+                {
+                    // Calculates the percent share then multiplies.
+                    stationShares[station] = (int)Math.Floor(((float)stationTotalSlots[station] / totalSlots) * candidates.Count);
+                    distributed += stationShares[station];
+                }
+
+                // Avoids the fair share problem where if there's two stations and one player neither gets one.
+                if (distributed < candidates.Count)
+                {
+                    var choice = _random.Pick(stations);
+                    stationShares[choice] += candidates.Count - distributed;
+                }
+
+                // Actual meat, goes through each station and shakes the tree until everyone has a job.
+                foreach (var station in stations)
+                {
+                    if (stationShares[station] == 0)
+                        continue;
+
+                    var slots = stationSlots[station];
+                    var allJobs = slots.Keys.ToList();
+                    _random.Shuffle(allJobs);
+                    // And iterates through all it's jobs in a random order until the count settles.
+                    // No, AFAIK it cannot be done any saner than this. I hate "shaking" collections as much
+                    // as you do but it's what seems to be the absolute best option here.
+                    // It doesn't seem to show up on the chart, perf-wise, anyway, so it's likely fine.
+                    int priorCount;
+                    do
+                    {
+                        priorCount = stationShares[station];
```

Comment:
this whole function is highkey unreadable. its commented well but i think it needs to be split up somehow

Reply:
Commented even more.